flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/MessageAcknowledingSourceBase.java

```diff
@@ -157,17 +179,35 @@ public void restoreState(SerializedCheckpointData[] state) throws Exception {
 
 	@Override
 	public void notifyCheckpointComplete(long checkpointId) throws Exception {
-		for (Iterator<Tuple2<Long, List<Id>>> iter = pendingCheckpoints.iterator(); iter.hasNext();) {
-			Tuple2<Long, List<Id>> checkpoint = iter.next();
-			long id = checkpoint.f0;
-			
-			if (id <= checkpointId) {
-				acknowledgeIDs(checkpoint.f1);
-				iter.remove();
-			}
-			else {
-				break;
+		if (!running) {
+			LOG.debug("notifyCheckpointComplete() called on closed source");
+			return;
+		}
+		
+		// only one commit operation must be in progress
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Committing Messages externally for checkpoint {}", checkpointId);
+		}
+		
+		List<Id> idsOfCheckpointedMessages = new LinkedList<Id>();
```

Comment:
The extra copying into a list only to generate a log statement seems like too much overhead.
Also, logging potentially millions of IDs the log (especially on INFO level) is probably a bit overkill...

I would log the IDs on TRACE level only. Also, rather than copy the individual lists into one lists, I would save the overhead and log one line per list.